File: algo/160. Intersection of Two Linked Lists.js
// @9 Linked List

/**
 * Definition for singly-linked list.
 * function ListNode(val) {
 *     this.val = val;
 *     this.next = null;
 * }
 */

/**
 * @param {ListNode} headA
 * @param {ListNode} headB
 * @return {ListNode}
 */
var getIntersectionNode = function(headA, headB) {
    if (headA === null || headB === null) {
        return null;
    }
    var cursor = headA;
    while(cursor.next !== null) {
        cursor = cursor.next;
    }
    var breakpoint = cursor;
    cursor.next = headB;
    var tortois = headA;
    var rabbit = headA;
    cursor = headA;
    while (rabbit !== null && rabbit.next !== null) {
        tortois = tortois.next;
        rabbit = rabbit.next.next;
        if (tortois === rabbit) {
            while (cursor !== tortois) {
                cursor = cursor.next;
                tortois = tortois.next;
            }
            breakpoint.next = null;
            return cursor;
        }
    }
    breakpoint.next = null;
    return null;
};

File: round3/160. Intersection of Two Linked Lists.cs
/**
 * Definition for singly-linked list.
 * public class ListNode {
 *     public int val;
 *     public ListNode next;
 *     public ListNode(int x) { val = x; }
 * }
 */
public class Solution {
    public ListNode GetIntersectionNode(ListNode headA, ListNode headB) {
        ListNode c1 = headA, c2 = headB;
        int cnt1 = 0, cnt2 = 0;
        while (c1 != null) {
            c1 = c1.next;
            cnt1 ++;
        }
        
        while (c2 != null) {
            c2 = c2.next;
            cnt2 ++;
        }
        
        c1 = headA;
        c2 = headB;
        
        while (cnt1 > cnt2) {
            c1 = c1.next;
            cnt1 --;
        }
        
        while (cnt1 < cnt2) {
            c2 = c2.next;
            cnt2 --;
        }
        
        while (c1 != c2 && c1 != null) {
            c1 = c1.next;
            c2 = c2.next;
        }
        
        return c1;
    }
}